Simplify key-based thread locals

joboet opened the pull request:
This PR simplifies key-based thread-locals by:
* unifying the macro expansion of `const` and non-`const` initializers
* reducing the amount of code in the expansion
* simply reallocating on recursive initialization instead of going through `LazyKeyInner`
* replacing `catch_unwind` with the shared `abort_on_dtor_unwind`

It does not change the initialization behaviour described in #110897.

joboet committed:
std: simplify key-based thread locals

cuviper commented:
r? @m-ou-se 

^ since you hopefully remember more about this code from #110897.

m-ou-se commented on library/std/src/sys/thread_local/os_local.rs:
```diff
@@ -85,78 +48,78 @@ pub macro thread_local_inner {
+            self.os.set(ptr as *mut u8);
+        }
+        if !old.is_null() {
+            // If the variable was recursively initialized, drop the old value.
+            // SAFETY: We cannot be inside a `LocalKey::with` scope, as the
+            // initializer has already returned and the next scope only starts
+            // after we return the pointer. Therefore, there can be no references
+            // to the old value.
+            drop(unsafe { Box::from_raw(old) });
```
I think it'd be good to swap this around (such that it behaves like OnceCell) in a future PR. (That PR will need an FCP and perhaps a crater run though.)

m-ou-se commented:
This looks great.

@bors r+

bors commented:
:pushpin: Commit 5fd4a84a8cf0059ad04cc3734419b3346a62934f has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se 5fd4a84a8cf0059ad04cc3734419b3346a62934f -->
<!-- homu: {"type":"Approved","sha":"5fd4a84a8cf0059ad04cc3734419b3346a62934f","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

m-ou-se commented:
@bors r-

joboet commented:
The failing UI tests only check internal perma-unstable implementations, so I think they are fine to remove.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #123339) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joboet commented:
r? libs

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit cef00ea07649327ab786a29670816b486a19d089 has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se cef00ea07649327ab786a29670816b486a19d089 -->
<!-- homu: {"type":"Approved","sha":"cef00ea07649327ab786a29670816b486a19d089","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #116123) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joboet committed:
std: clean up the TLS implementation

joboet commented:
Rebased to resolve the conflict with the native TLS rewrite. I've also added a second commit that makes the `#[allow(dead_code)]` more local and removed the unnecessary imports that are now detected.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9222029675/job/25372344745) [(plain)](https://github.com/rust-lang/rust/commit/90ef5979b68ee28dd3ef5fd62ec59dbb7aa390d9/checks/25372344745/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9222239420/job/25373007050) [(plain)](https://github.com/rust-lang/rust/commit/5a7e50e37eef3f27997be6ae2132d47098c332cf/checks/25373007050/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling core v0.0.0 (/checkout/library/core)
error: function `register_dtor` is never used
  --> library/std/src/sys/pal/unix/thread_local_dtor.rs:26:15
   |
26 | pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`

```

</details>

m-ou-se commented:
@bors r+

bors commented:
:pushpin: Commit 0e7e75ebca2ee9cc9a6c019110797250094908f5 has been approved by `m-ou-se`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=m-ou-se 0e7e75ebca2ee9cc9a6c019110797250094908f5 -->
<!-- homu: {"type":"Approved","sha":"0e7e75ebca2ee9cc9a6c019110797250094908f5","approver":"m-ou-se","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0e7e75ebca2ee9cc9a6c019110797250094908f5 with merge 9e297bf54d31eb3b30067208ff9af4416945a2ed...
<!-- homu: {"type":"BuildStarted","head_sha":"0e7e75ebca2ee9cc9a6c019110797250094908f5","merge_sha":"9e297bf54d31eb3b30067208ff9af4416945a2ed"} -->
